### ISSUE TITLE
anything is showable as text/csv

### ISSUE THREAD

**Issue by SimonDanisch** [JuliaLang/julia#31037]
Any Julia type is showable as `text/csv` by default, which makes it hard to actually implement a display that supports specifically the text/csv mime.
This happens because of this method:
https://github.com/JuliaLang/julia/blob/master/stdlib/DelimitedFiles/src/DelimitedFiles.jl#L828
What makes it worse is, that not even all types are actually showable:

```Julia
struct Test end
Base.showable(MIME"text/csv"(), Test())
Base.show(stdout, MIME"text/csv"(), Test())
MethodError: no method matching iterate(::Test)
Closest candidates are:
  iterate(!Matched::Core.SimpleVector) at essentials.jl:568
  iterate(!Matched::Core.SimpleVector, !Matched::Any) at essentials.jl:568
  iterate(!Matched::ExponentialBackOff) at error.jl:199
  ...
#writedlm#14(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Base.TTY, ::Test, ::Char) at DelimitedFiles.jl:779
writedlm(::Base.TTY, ::Test, ::Char) at DelimitedFiles.jl:776
show(::Base.TTY, ::MIME{Symbol("text/csv")}, ::Test) at DelimitedFiles.jl:828
top-level scope at none:0
```

**Comment by nalimilan** [comment]
That line should probably be restricted to `AbstractArray`. Also I think @JeffBezanson mentioned in another issue that the CSV printing code is simple enough that it could be included in Base, which would avoid what is technically type piracy.

**Comment by JeffBezanson** [comment]
Yes, this was found by #30945. I agree it should be restricted to AbstractArray (maybe just AbstractVector and AbstractMatrix). However, the code deliberately implements and tests support for arbitrary iterators of iterators, so technically it's a loss of functionality. I highly doubt anybody is using it though. Would be good to think of another API for that.

**Comment by SimonDanisch** [comment]
I just noticed, that there is `isiterable`, so we could define at least:
```Julia
showable(::MIME"text/csv", x::T) where T = Base.isiterable(T) || T <: AbstractVector || T <: AbstractMatrix
```

**Comment by JeffBezanson** [comment]
I don't think that's good enough. It needs to be an iterator of iterators of elements that can be reasonably printed as CSV, which is a pretty marginal use case for such an invasive method definition.
